### DISCUSSION TITLE
Is possible to save only the Layers?

### DISCUSSION THREAD

**Discussion by oscarshaitan** [hm21/pro_image_editor#268] [Q&A] [answered]
I want to save only the layers. with a transparency background
I used the example/lib/pages/movable_background_image.dart
and on onImageEditingStarted I remove the layer, but that requires a fine tune in the stickerEditorConfigs.initWidth and I manage to land a pretty good case but changes in window , screens affect that calculation.

also I tried to clear the editorImage  but that didnt work (also I think that is a bad awful idea)

My last hope is to save the history and rebuild the  ProImageEditor.memory with a transparent background  and in the onAfterViewInit add all the history. But I think this is not optimal
Update: the 2 editors strategy and passing the history is working but still think is too overkill

any suggestions?

**Comment by hm21** [comment]
Could you share how you are currently using the `ProImageEditor` widget in your code? I'd like to review the `configs` to better understand your setup. Depending on the sub-editors you're using (such as the Sticker and Text editors), it might be possible to place the image outside of the editor. However, I’d prefer to examine the configuration first.

If you prefer the two-editor strategy, it should work as intended. If you haven’t already done so, ensure that the `generateImageInBackground` is disabled in the first editor. Keeping it enabled would be redundant if the image from that editor isn't being used. Additionally, you should set the `processorMode` to `ProcessorMode.minimum`, as it's unnecessary for the editor to set up multiple threads in this case.

```dart
configs: ProImageEditorConfigs(
  imageGenerationConfigs: const ImageGenerationConfigs(
    generateImageInBackground: false,
    processorConfigs: ProcessorConfigs(
      processorMode: ProcessorMode.minimum
    )
  )
)
```
There is also a function, [generateHighQualityImage](https://github.com/hm21/pro_image_editor/blob/1858aac8f0a37ad186bb2f8a0002540376ed943f/lib/utils/content_recorder.dart/utils/generate_high_quality_image.dart#L27-L59), that you could use. This function generates the image in a separate thread (web worker/isolated), but it requires you to provide the raw image data. To obtain the raw image data, you'll need to wrap all layers inside a `RepaintBoundary` and access it using a `GlobalKey`.

Anyway, it would be best if you shared the code for your `ProImageEditor` widget with the configs. That way, I can consider which is the best approach for your needs.

**Reply by oscarshaitan** [reply]
ill include those configurations to the first editor thanks.

**Comment by oscarshaitan** [comment]
I started by placing the Image as a Sticker but that give me issues with the size of the sticker vs the size of the background. 
then I move to 2 editors 1 to create the layers get the history pass that to a second editor with a transparent background (similar to what you have on the EditorImage with no results, also changing the parameter byteArray no results.


This is the Editor I use to capture the layers
![Screenshot 2024-11-25 at 10 07 32](https://github.com/user-attachments/assets/02d1396c-a38c-4034-90c4-cbd21412573f)


And this the second one to save the layers without the background
![Screenshot 2024-11-25 at 10 07 54](https://github.com/user-attachments/assets/54e350bc-df48-4c37-8c25-3f3c456de10e)

**Reply by hm21** [reply]
Apologies for the delayed response.

An alternative to the two-editor solution could involve replacing the image directly when the user presses the "Done" button. To implement this, you can enhance the Done button in your app bar by adding the following code:

```dart
/// import 'dart:ui' as ui;

Size imageSize = editor.sizesManager.originalImageSize ?? editor.sizesManager.decodedImageSize;

double width = imageSize.width;
double height = imageSize.height;

final recorder = ui.PictureRecorder();
final canvas = Canvas(recorder, Rect.fromLTWH(0, 0, width.toDouble(), height.toDouble()));
final paint = Paint()..color = Colors.transparent;
canvas.drawRect(
    Rect.fromLTWH(0.0, 0.0, width.toDouble(), height.toDouble()),
    paint);

final picture = recorder.endRecording();
final img = await picture.toImage(width.toInt(), height.toInt());
final pngBytes = await img.toByteData(format: ui.ImageByteFormat.png);

editor.editorImage = EditorImage(
  byteArray: pngBytes!.buffer.asUint8List(),
);
await editorKey.currentState!.decodeImage();

/// Force to rebuild everything
int pos = editor.stateManager.position;
EditorStateHistory oldHistory = editor.stateManager.stateHistory[pos];

editor.stateManager.stateHistory[pos] = EditorStateHistory(
  layers: oldHistory.layers,
  transformConfigs: oldHistory.transformConfigs,
  blur: oldHistory.blur,
  filters: [...oldHistory.filters],
  tuneAdjustments: [...oldHistory.tuneAdjustments],
);

editor.doneEditing();
```

For the `imageGenerationConfigs`, you can configure it as follows for optimized performance:

```dart
imageGenerationConfigs: const ImageGenerationConfigs(
  captureOnlyDrawingBounds: false,
  generateImageInBackground: false,
  outputFormat: OutputFormat.png,
  processorConfigs: ProcessorConfigs(
    processorMode: ProcessorMode.minimum,
  ),
),
```

**Reply by oscarshaitan** [reply]
Dont worry thanks.
the package is so good that now we are going to expand the feature. I'm not longer going to save the images as Layers. Instead I'm going to serialize the class Layers so I can save it in the DB and be able to edit later. and for saving they will save the enable layers.

Thanks for your help and once more awesome package!!!

**Comment by oscarshaitan** [comment, marked as answer]
So far having 2 editors is the solution for my case
